### PR TITLE
fix(curriculum): make challenge Helmet 4 Compatible

### PR DESCRIPTION
**Pull request by SethFalco**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

**Blocked by: https://github.com/freeCodeCamp/boilerplate-project-secure-real-time-multiplayer-game/pull/10**

This relates to the following issue and pull request.
Issue: https://github.com/freeCodeCamp/freeCodeCamp/issues/40476
Pull Request: https://github.com/freeCodeCamp/freeCodeCamp/pull/40900

> `helmet.hidePoweredBy` removes the `X-Powered-By` header, which is set by default in some frameworks (like Express). Removing the header offers very limited security benefits (see [this discussion](https://github.com/expressjs/express/pull/2813#issuecomment-159270428)) and is mostly removed to save bandwidth.
> \- https://helmetjs.github.io/

On the same page, it's recommended not to use `app.use(helmet.hidePoweredBy())` and instead opt for `app.disable("x-powered-by")`, anyway. 

---

I updated the test to only verify that the `x-powered-by` header isn't equal to `Express`, this logic is stolen from a challenge before the projects.

https://github.com/freeCodeCamp/freeCodeCamp/blob/main/curriculum/challenges/english/09-information-security/information-security-with-helmetjs/hide-potentially-dangerous-information-using-helmet.hidepoweredby.md#--hints--

I removed the instructions suggesting that `helmet@^3.21.3` is required to complete the user stories.

### CODE REVIEW

**Comment by nhcarrigan** [comment]
Hey @SethFalco are you still working on this draft? Or is it ready for review?

**Comment by SethFalco** [comment]
> Hey @SethFalco are you still working on this draft? Or is it ready for review?

Sorry about that, I've been busy, so I haven't been able to finish this yet.

> Converting this to a draft, I thought I had tested this properly, but seems I made a mistake.
> app.use(helmet.xssFilter()); now disables XSS filtering. This results in the x-xss-protection header being set to 0.
> 
> I think this test should be replaced with using an appropriate Content Security Policy instead that doesn't allow unsafe-inline in JavaScript at least. I'll update this PR later.
> \- https://github.com/freeCodeCamp/boilerplate-project-secure-real-time-multiplayer-game/pull/10#issuecomment-803071398

It's in a draft state because of the above, I can try to come back to this on the weekend and get it in a reviewable state.

**Comment by nhcarrigan** [comment]
Sounds good - thanks for the update 🙂

**Comment by SethFalco** [comment]
Applied the change now, I referenced another challenge to check the CSP.
I might update it later to more specifically check that `unsafe-inline` is not used, rather than checking if it's set to `safe`.

Just not sure should the task should specify to use `self`, or if the test just check if it doesn't contain `unsafe-inline`.
If we're going to check if it's not `unsafe-inline`, it might be a good idea to depend on a package that will actually parse the content-security policy properly. It's certainly possible through regex as well, but across challenges using a CSP parser would be cleaner, I think.

I'll check why the build is failing in the evening.

**Comment by nhcarrigan** [comment]
Build error is unrelated to your PR - I'll try re-running the action to see if it passes.

/cc @ojeytonwilliams same caching issue on this Cypress run.

**Comment by ojeytonwilliams** [comment]
Thanks for the heads up @nhcarrigan.  Still figuring this one out, but I tried merging in main in the hope that that would bring the webpack-runtimes into alignment.

@SethFalco sorry about this, hopefully the merge commit will sort this out.  If not, we can still merge, even if that test fails.

**Review by raisedadead — APPROVED** [review]

**Review by moT01 — COMMENTED** [review]
I am going to vote to not add this and the related PR.

We've had several discussions around helmet, it keeps coming up - [here's one](https://github.com/freeCodeCamp/freeCodeCamp/issues/39190) where we settled on keeping at helmet 3.

I don't mind a few extra tests that would allow for the use of helmet 4x. The XSS test changed here will likely cause all previously submitted projects to break. We could adjust it to accept the old way or how this PR has it to account for both. That would really be the only thing to change here to keep it backward compatible. So it wouldn't be difficult for the time being.

What happens when helmet 5x changes more things. Keep updating things, I suppose. The main learning goal here is more about headers and security than the specific methods or versions of a package used.

Regarding the boilerplate PR - I will just leave a comment here:

Helmet removed the `nocache` method into a separate package (which is added in the boilerplate PR). The curriculum doesn't mention that anywhere, users would likely try and use the method in the curriculum and it wouldn't work. Which really illustrates the root problems in making adjustments like this. The curriculum is pinned to 3x, the other projects and boilerplates use 3x (as far as I know). I dunno, maybe I'm being stubborn.

To reiterate, I wouldn't mind adding this PR if the tests accept the 3x or 4x methods. I think the boilerplate should be left alone though.

**Comment by SethFalco** [comment]
Just dropping my opinion on this. Please note I'm just a contributor, I may've misunderstood the objectives of freeCodeCamp or contribution guidelines.

TL;DR: I think _ideally_ updatable educational material should take advantage of this and update library versions as well, _especially_ major versions. My opinion is that the real solution is to make this a draft while PRs are made to get the curriculum up-to-date, rather than settle on leaving it as Helmet 3.

---

> We could adjust it to accept the old way or how this PR has it to account for both. That would really be the only thing to change here to keep it backward compatible.

This didn't occur to me, I agree making the tests compatible with version 3 and 4 is a good idea.
I'll try to update the PR to do that, just have to think about what's the best approach.

> What happens when helmet 5x changes more things. Keep updating things, I suppose.

Ideally, I do think it'd be nice if new students were taught the up-to-date versions of software where possible. I recognize with how rapidly software changes this can prove difficult, but it's certainly an ideal.

> The main learning goal here is more about headers and security than the specific methods or versions of a package used.

If the main goal is to learn about headers and security, I feel it would've made more sense for users to just set the headers directly instead of using Helmet? Fewer dependencies, and get to learn about headers more directly.

For example, since I opted to use Helmet 4, and just set the headers myself.

```js
app.use((req, res, next) => {
  res.set('x-powered-by', 'PHP 7.4.3');
  res.set('x-xss-protection', '1; mode=block')
  next();
})
```

> Helmet removed the nocache method into a separate package (which is added in the boilerplate PR). The curriculum doesn't mention that anywhere

Agreed, I believe the ideal solution would be to turn this PR into a draft, and I can separately make a PR which includes this information in the curriculum.

> I dunno, maybe I'm being stubborn.

On the flip side, maybe I'm being petty.

---

As a new contributor, I wouldn't mind updating the curriculum to teach version 4, and all projects to use version 4 with version 3 compatibility, it might take a while though. It may be a nice way to try to get comfortable with the project, so I contribute more in future.

I'd just like to see people be able to use Helmet 4 immediately when they take this knowledge to personal projects, or to the workplace. Not realize some of the practices they were taught were outdated and have to read the through Helmet documentation to see what changed, and maybe MDN Web Docs, and issues on the Helmet repo to see why.

Helmet isn't the only example either, I'd love to see the D3 tutorials updated from v4 to v6 too, I believe that one wouldn't require changing tests, just how some asynchronous methods are called.

**Comment by ShaunSHamilton** [comment]
@SethFalco Thank you, for sharing. You are spot-on with a lot of your points. I will just add this, and no more, as a PR is not the place to have too much discussion on this topic:

With version 7 of the curriculum being the main focus of the freeCodeCamp team's energy, the current curriculum is not receiving massive changes; typically, bug fixes are done, because they can be frustrating to learners, and it makes sense to do whilst the curriculum is still active (online). This is why the main suggestion for the mentioned issue has been to quickfix the bug - we are slowly learning to freeze the versions of the tools we use.

Absolutely, Campers should be taught up-to-date tech. However, as tech moves so fast, and the core fCC team is so small, there is little-to-no scope to continually update a curriculum which is scheduled for obsolescence. That is the issue in its entirety - scope/man-power.

Now, I think a complete overhaul of this section is unlikely to do much good, because:
a) It will take a lot of time, leaving the issue an issue for too long
b) It could very well only be on the curriculum fore-front for a short amount of time
However, I appreciate your willingness to contribute in this way. So, I will leave this to be decided by the others, as they know the curriculum plans much better.

**Comment by huyenltnguyen** [comment]
I am also leaning toward the preserved side. 

I'm not sure how much effort it would take to make the upgrade backward compatible, but it would require a lot of testing and reviewing effort from the maintainers, as well as informing forum moderators and campers in case any issues arise. All of these make the change seem costly to me, especially at the moment, when the new curriculum is the main focus.

@SethFalco We treasure every minute of contributors' time and we really appreciate the improvements you have been making across fCC repos. I think it would be more beneficial if we could have your contributions invested into the development of the new curriculum instead. 🙂

**Comment by raisedadead** [comment]
Since the consensus was to not add the change and instead ensure this is covered in the upcoming curriculum going to close this.
